### ISSUE TITLE
PHP 7 Typehintings Tracker

### ISSUE THREAD

**Issue by massimilianobraglia** [ruflin/Elastica#1548]
In this issue I will update the work in progress.

Namespaces:
- [x] `Elastica\Aggregation` (https://github.com/ruflin/Elastica/pull/1549)
- [x] `Elastica\Bulk` (https://github.com/ruflin/Elastica/pull/1550)
- [x] `Elastica\Cluster` (https://github.com/ruflin/Elastica/pull/1551)
- [x] `Elastica\Connection` (https://github.com/ruflin/Elastica/pull/1560)
- [x] `Elastica\Exception` (https://github.com/ruflin/Elastica/pull/1568)
- [x] `Elastica\Index` (https://github.com/ruflin/Elastica/pull/1569)
- [x] `Elastica\Multi` (https://github.com/ruflin/Elastica/pull/1576)
- [x] `Elastica\Node` (https://github.com/ruflin/Elastica/pull/1577)
- [x] `Elastica\Processor` (https://github.com/ruflin/Elastica/pull/1579)
- [x] `Elastica\Query` (https://github.com/ruflin/Elastica/pull/1582)
- [x] `Elastica\QueryBuilder` (https://github.com/ruflin/Elastica/pull/1583)
- [x] `Elastica\Rescore` (https://github.com/ruflin/Elastica/pull/1566)
- [x] `Elastica\ResultSet` (https://github.com/ruflin/Elastica/pull/1565)
- [x] `Elastica\Script` (https://github.com/ruflin/Elastica/pull/1564)
- [x] `Elastica\Suggest` (https://github.com/ruflin/Elastica/pull/1562)
- [x] `Elastica\Transport` (https://github.com/ruflin/Elastica/pull/1558)
- [x] `Elastica\Type` (https://github.com/ruflin/Elastica/pull/1557)
- [ ] `Elastica` base namespace

Also tests will be updated.

**Comment by osavchenko** [comment]
May I help with this feature?

**Comment by p365labs** [comment]
@massimilianobraglia maybe you and @osavchenko could split the work in two and divide the effort ?
what do u think ?

**Comment by massimilianobraglia** [comment]
That's a good idea @p365labs! Thank you, @osavchenko 

I was processing `Connection` namespace, so you can start with the others uncheck

**Comment by osavchenko** [comment]
@massimilianobraglia I'll start from the end of the list and we will meet each other in the middle ;)

**Comment by massimilianobraglia** [comment]
@osavchenko fine for me! 😄

**Comment by osavchenko** [comment]
Should we use `strict_types`?

**Comment by massimilianobraglia** [comment]
From my POV using `strict_types` is necessary. The only problem is that declaring strict types on every file will cause a huge BC, and that is not something a library wants to 😄 

@ruflin and @p365labs my proposal about strict types is to release those in the next major release. What do think about it?

**Comment by ruflin** [comment]
Agree that we should only introduce it in a next major. After all these PR's went in we should probably branch of 6.x and then we can start with strict_type. What do you expect will be the impact on our users. Is this something already commonly used? Is PHP planning to enable something like this by default?

**Comment by massimilianobraglia** [comment]
No, I don't think that PHP is planning to introduce this by default. They didn't do it as of 7.3.
Anyway, that `declare` can be set as a rule of PHP-CS-Fixer, so don't worry about setting it manually.

Strict types are already in use by some Symfony bundles, Doctrine's libraries and so on

**Comment by ruflin** [comment]
Thanks for the details. Mainly asking to better understand for which and how many of the Elastica users this could become a challenge.

**Comment by p365labs** [comment]
@ruflin  @massimilianobraglia sorry I missed the last message, I agree on the idea of ```having strict_types``` on the next major

**Comment by osavchenko** [comment]
I'm not sure, probably - not

**Comment by massimilianobraglia** [comment]
@ruflin it was just an idea:
- new PR with fixer rule that introduce the `declare(strict_type);` statement
- run the linter
- execute test
- from time to time rebasing it from master and run again test to check that the strict_typeness won't broke things
- when the major comes then we can merge it

**Comment by ruflin** [comment]
Got it. Probably good as a one time thing to check. I was kind of hoping we could enable it somehow / somewhere to for the test runs.

**Comment by massimilianobraglia** [comment]
@ruflin and @p365labs sorry about the delay in completing this one. It's a period very full :(

Maybe can you start reviewing what's been done?

Meanwhile I'm trying to complete the last point.

**Comment by ruflin** [comment]
@massimilianobraglia Sorry for the delay on reviews here on my end. I was kind of waiting on our decision on the defaults (good excuse ? :-) ). I like your most recent proposal.

**Comment by massimilianobraglia** [comment]
No problem about the delay 😄 I know that feeling!

**Comment by p365labs** [comment]
Hi @ruflin yep! I totally agree it's a good compromise.

**Comment by massimilianobraglia** [comment]
Ok then, I'll update my PRs ASAP 😄

**Comment by osavchenko** [comment]
Guys, I have changed my work and little bit busy last days, so, I remember about PR's and will update them soon

**Comment by p365labs** [comment]
@massimilianobraglia & @osavchenko  well done!!! Thank you very much. It has been an impressive work!

**Comment by massimilianobraglia** [comment]
Thank you, @p365labs! Now that all PRs are merged I'll work on the Elastica base namespace!

**Comment by osavchenko** [comment]
Thank you all! It was an interesting experience and I'm glad to work with you ;)
